types/three/examples/jsm/nodes/shadernode/ShaderNode.d.ts

```diff
@@ -1,49 +1,59 @@
-import { AnyObject, NodeTypeOption, SwizzleOption } from "../core/constants.js";
 import ConstNode from "../core/ConstNode.js";
 import Node from "../core/Node.js";
 import NodeBuilder from "../core/NodeBuilder.js";
+import StackNode from "../core/StackNode.js";
 
 export interface NodeElements {
+    toGlobal: (node: Node) => Node;
+
     append: typeof append;
 
-    color: typeof color;
-    float: typeof float;
-    int: typeof int;
-    uint: typeof uint;
-    bool: typeof bool;
-    vec2: typeof vec2;
-    ivec2: typeof ivec2;
-    uvec2: typeof uvec2;
-    bvec2: typeof bvec2;
-    vec3: typeof vec3;
-    ivec3: typeof ivec3;
-    uvec3: typeof uvec3;
-    bvec3: typeof bvec3;
-    vec4: typeof vec4;
-    ivec4: typeof ivec4;
-    uvec4: typeof uvec4;
-    bvec4: typeof bvec4;
-    mat2: typeof mat2;
-    imat2: typeof imat2;
-    umat2: typeof umat2;
-    bmat2: typeof bmat2;
-    mat3: typeof mat3;
-    imat3: typeof imat3;
-    umat3: typeof umat3;
-    bmat3: typeof bmat3;
-    mat4: typeof mat4;
-    imat4: typeof imat4;
-    umat4: typeof umat4;
-    bmat4: typeof bmat4;
-    string: typeof string;
-    arrayBuffer: typeof arrayBuffer;
+    toColor: typeof color;
+    toFloat: typeof float;
+    toInt: typeof int;
+    toUint: typeof uint;
+    toBool: typeof bool;
+    toVec2: typeof vec2;
+    toIvec2: typeof ivec2;
+    toUvec2: typeof uvec2;
+    toBvec2: typeof bvec2;
+    toVec3: typeof vec3;
+    toIvec3: typeof ivec3;
+    toUvec3: typeof uvec3;
+    toBvec3: typeof bvec3;
+    toVec4: typeof vec4;
+    toIvec4: typeof ivec4;
+    toUvec4: typeof uvec4;
+    toBvec4: typeof bvec4;
+    toMat2: typeof mat2;
+    toImat2: typeof imat2;
+    toUmat2: typeof umat2;
+    toBmat2: typeof bmat2;
+    toMat3: typeof mat3;
+    toImat3: typeof imat3;
+    toUmat3: typeof umat3;
+    toBmat3: typeof bmat3;
+    toMat4: typeof mat4;
+    toImat4: typeof imat4;
+    toUmat4: typeof umat4;
+    toBmat4: typeof bmat4;
 
     element: typeof element;
     convert: typeof convert;
 }
 
 export function addNodeElement(name: string, nodeElement: unknown): void;
 
+export type SwizzleCharacter = "x" | "y" | "z" | "w" | "r" | "g" | "b" | "a" | "s" | "t" | "p" | "q";
+
+export type SwizzleOption = Exclude<
+    | `${SwizzleCharacter}`
+    | `${SwizzleCharacter}${SwizzleCharacter}`
+    | `${SwizzleCharacter}${SwizzleCharacter}${SwizzleCharacter}`
+    | `${SwizzleCharacter}${SwizzleCharacter}${SwizzleCharacter}${SwizzleCharacter}`,
+    "abs" | "sqrt"
+>;
+
```

Comment:
Hi! Is there a more performant way to type something like this? I believe this is making type checking working with TSL / nodes very slow.

I bisected my extensions, restarted TS server, etc to no avail but removing this type seems to speed things up quite a bit.

<img width="365" alt="Screenshot 2024-06-30 at 14 14 37" src="https://github.com/DefinitelyTyped/DefinitelyTyped/assets/4625978/a771d0b4-d047-4d50-97c6-d2bc451f3268">

This UI is much less useful with all the swizzle permutations now too :(

Reply:
I'm not sure, can you create an issue on https://github.com/three-types/three-ts-types so this doesn't get lost? We could try manually listing out all the swizzable options, but my guess is that the slowness comes from the large number of swizzable options in the discriminated union, not the cost of computing what they are. Perhaps there are some swizzable options that are listed that don't actually make sense (e.g., `aaaa`?)?